torch/_inductor/fx_passes/group_batch_fusion.py

```diff
@@ -740,46 +741,100 @@ def __init__(self, **kwargs):
 
 
+class _OrderedSet:
+    def __init__(self, param=None):
+        if param:
+            self.rep = OrderedDict({k: None for k in param})
+        else:
+            self.rep = OrderedDict()
+
+    def __contains__(self, o):
+        return o in self.rep
+
+    def __len__(self):
+        return self.rep.__len__()
+
+    def append(self, o):
+        self.rep[o] = None
+
+    def __iter__(self):
+        return self.rep.keys().__iter__()
+
+
 def find_independent_subset_greedy(
-    node_list: List[torch.fx.Node],
+    node_list: abc.Iterable[torch.fx.Node],
     graph_search_options: Dict[str, Any],
-) -> Iterator[List[torch.fx.Node]]:
+) -> Iterator[abc.Iterable[torch.fx.Node]]:
     """
-    Return a list of subset from node_list, all nodes in each subset are independent with each other and can be fused together.
-    The type of subset is list, so we can preserve node's order and benefit from split-cat elimination in later pass.
+    Returns a list of subsets of `node_list` where no element in the subset
+    depends on any other element in the subset. This results in a set of
+    independent nodes which can be fused together.
+
+    The order of `node_list` is preserved within each subset so we can benefit
+    from split-cat elimination in later passes.
+
+    graph_search_options:
+      - min_fuse_set_size: Minimum size of the subset to consider. Subsets below
+        this size will be ignored.
+      - max_fuse_set_size: Maximum size of the subset to consider. Subsets will
+        be broken to be at most this size.
     """
-    visited_node_set: Set[torch.fx.Node] = set()
-    dep_set: Set[torch.fx.Node] = set()
 
-    def find_dependent_nodes(src_node, cur_node):
-        for input_node in cur_node.all_input_nodes:
-            if input_node in node_list:
-                dep_set.add(input_node)
+    # Compute all the children of `node` which are members of
+    # `interesting_nodes`.
+    def find_dependent_nodes(node, interesting_nodes):
+        visited_node_set: Set[torch.fx.Node] = {node}
+        dep_set: Set[torch.fx.Node] = set()
+
+        work = [node]
+        while work:
+            node = work.pop()
+            for input_node in node.all_input_nodes:
+                if input_node in interesting_nodes:
+                    dep_set.add(input_node)
 
-            if input_node not in visited_node_set:
-                visited_node_set.add(input_node)
-                find_dependent_nodes(src_node, input_node)
+                if input_node not in visited_node_set:
+                    visited_node_set.add(input_node)
+                    work.append(input_node)
 
-    while len(node_list) > 0:
+        return dep_set
+
+    min_fuse_set_size = graph_search_options["min_fuse_set_size"]
+    max_fuse_set_size = graph_search_options["max_fuse_set_size"]
+
+    # Cache for the computed dep_sets when we know we're going to need it again.
+    cache: Dict[torch.fx.Node, Set[torch.fx.Node]] = {}
```

Comment:
Isn't this cache invalid if any fusions happen?

Reply:
I'm not done with this PR yet (it's still in draft).  But yes - do we do that? I would imagine that changing the graph during iteration would cause problems for the original algorithm as well.

Reply:
This coroutine yields `subset` back into `apply_group_batch_fusion` which may alter the graph. The current algorithm doesn't carry any state over to the next iteration of the while loop, so I don't see why it would be an issue there.

Reply:
Wouldn't it be a problem if a node in `node_list` was removed from the graph? The old algorithm would still consider that node in its check even though the node is no longer in the graph. If we're just fusing nodes from `subset` that are being returned then I guess that wouldn't be an issue because they've already been removed from `node_list`.

Anyway - you're right that the way the call is being used would cause a problem with the cache - I'll remove it before I publish. Good catch.

Reply:
>  If we're just fusing nodes from subset that are being returned then I guess that wouldn't be an issue because they've already been removed from node_list.

This is correct, it only considers `subset`